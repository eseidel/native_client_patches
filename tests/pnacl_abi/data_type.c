/*
 * tests whether our data type are lowered as expected (ILP32)
 * TODO: sizes of size_sb and size_ue are not quite right.
 */

#include <setjmp.h>
#include <stdarg.h>
#include <stddef.h>
#include <unwind.h>

typedef struct _Unwind_Exception UE;

int size_c = sizeof(char);
int size_i = sizeof(int);
int size_l = sizeof(long);

int size_p = sizeof(void*);

int size_f = sizeof(float);
int size_d = sizeof(double);

/* A 24 bytes buffer to accomodate all three architectures */
int size_vl = sizeof(va_list);

/* A 1024 byte buffer to accomodate future architectures */
int size_jb = sizeof(jmp_buf);
/* this should just be 4 words but we needed to hack it */
/* c.f. http://code.google.com/p/nativeclient/issues/detail?id=1107 */
int size_ue = sizeof(UE);


#define SET_ALIGNMENT(T, name) \
  typedef struct { char dummy; T x; } AlignStruct_ ## name; \
  int align_ ## name = offsetof(AlignStruct_ ## name, x)

SET_ALIGNMENT(char, c);
SET_ALIGNMENT(int, i);
SET_ALIGNMENT(long, l);

SET_ALIGNMENT(void*, p);

SET_ALIGNMENT(float, f);
SET_ALIGNMENT(double, d);

SET_ALIGNMENT(va_list, vl);

SET_ALIGNMENT(jmp_buf, jb);

SET_ALIGNMENT(UE, ue);
