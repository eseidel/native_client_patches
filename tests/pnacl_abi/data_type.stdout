target triple = "armv7-none-linux-gnueabi"

@size_c = global i32 1
@size_i = global i32 4
@size_l = global i32 4
@size_p = global i32 4
@size_f = global i32 4
@size_d = global i32 8
@size_vl = global i32 24
@size_jb = global i32 1024
@size_ue = global i32 56
@align_c = global i32 1
@align_i = global i32 4
@align_l = global i32 4
@align_p = global i32 4
@align_f = global i32 4
@align_d = global i32 8
@align_vl = global i32 8
@align_jb = global i32 8
@align_ue = global i32 8
